SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AutodiffClosureSpecialization.swift

```diff
@@ -0,0 +1,728 @@
+/// ``` swift
+/// @differentiable(reverse)
+/// func foo(_ x: Float) -> Float {
+///     return sin(x)
+/// }
+///
+/// //============== Before optimization ==============//
+/// sil @pb_foo: $(Float, (Float) -> Float) -> Float {
+/// bb0(%0: $Float, %1: $(Float) -> Float):
+///     %2 = apply %1(%0): $Float
+///     return %2: $Float
+/// }
+///
+/// sil @vjp_foo: $(Float) -> (Float, (Float) -> Float) {
+/// bb0(%0: $Float):
+///     %1 = apply @sin(%0): $Float
+///     %2 = partial_apply @pb_sin(%0): $(Float) -> Float
+///     %pb_foo = partial_apply @pb_foo(%2): $(Float, (Float) -> Float) -> Float
+///     return (%1, %pb_foo)
+/// }
+///
+/// //============== After optimization ==============//
+/// sil @specialized_pb_foo: $(Float, Float) -> Float {
+/// bb0(%0: $Float, %1: $Float):
+///     %2 = partial_apply @pb_sin(%1): $(Float) -> Float
+///     %3 = apply %2(%0): $Float
+///     return %3: $Float
+/// }
+///
+/// sil @vjp_foo: $(Float) -> (Float, (Float) -> Float) {
+/// bb0(%0: $Float):
+///     %1 = apply @sin(%0): $Float
+///     %pb_foo = partial_apply @specialized_pb_foo(%0): $(Float, Float) -> Float
+///     return (%1, %pb_foo)
+/// }
+/// ```
+
+import SIL
+
+private let verbose = false
+
+private func log(_ message: @autoclosure () -> String) {
+  if verbose {
+    print("### \(message())")
+  }
+}
+
+// =========== Entry point =========== //
+let autodiffClosureSpecialization = FunctionPass(name: "autodiff-closure-specialize") {
+  (function: Function, context: FunctionPassContext) in
+
+  // TODO: Don't optimize functions that are marked with the opt.never
+  // attribute.
+
+  // TODO: If F is an external declaration, there is nothing to specialize.
+
+  // If this is not an Autodiff VJP for a differentiable function, exit early.
+  if !function.isAutodiffVJP {
+    return
+  }
+
+  // TODO: Eliminate dead closures, return.
+
+  // TODO: Invalidate everything since we delete calls as well as add new
+  // calls and branches.
+}
+
+// =========== AutodiffClosureSpecializer ========== //
+
+struct AutodiffClosureSpecializer {
+  typealias PropagatedClosure = SingleValueInstruction
+  let specializationLevelLimit = 2
+  let context: FunctionPassContext
+
+  init(_ context: FunctionPassContext) {
+    self.context = context
+  }
+
+  func gatherCallSites(_ caller: Function) -> Stack<CallSiteDescriptor> {
+    // We should not look at reabstraction closures twice who we
+    // ultimately ended up using as an argument that we specialize on.
+    var usedReabstractionClosures = InstructionSet(self.context)
+    defer {
+      usedReabstractionClosures.deinitialize()
+    }
+
+    var paiToCallSiteDesc: OrderedDict<PartialApplyInst, CallSiteDescriptor> = OrderedDict()
+
+    for bb in caller.blocks {
+      for inst in bb.instructions {
+        // If this is not a closure that we support specializing, skip it.
+        if !inst.isSupportedClosure {
```

Comment:
I like the change you're suggesting. However, unless I'm misunderstanding something, it would just move the core work elsewhere, no?

The work we do for a single closure instruction is, go over all its uses in the function. This is the part that will now live in a separate function. Do you think we need to break this up further though?